src/funnel.ts

```diff
@@ -0,0 +1,280 @@
+   *
+   * @param args - The args are defined by the `reducer` function.
+   */
+
+  readonly call: (...args: Args) => void;
+
+  /**
+   * Resets the funnel to it's initial state. Any calls made since the last
+   * invocation will be discarded.
+   */
+  readonly cancel: () => void;
+
+  /**
+   * Triggers an invocation regardless of the current state of the funnel.
+   * Like any other invocation, The funnel will also be reset to it's initial
+   * state afterwards.
+   */
+  readonly flush: () => void;
+
+  /**
+   * The funnel is in it's initial state (there are no active timeouts).
+   */
+  readonly isIdle: boolean;
+};
+
+/**
+ * Creates a funnel that controls the timing and execution of `callback`. Its
+ * main purpose is to manage multiple consecutive (usually fast-paced) calls,
+ * reshaping them according to a defined batching strategy and timing policy.
+ * This is useful when handling uncontrolled call rates, such as DOM events or
+ * network traffic. It can implement strategies like debouncing, throttling,
+ * batching, and more.
+ *
+ * An optional `reducer` function can be provided to allow passing data to the
+ * callback via calls to `call` (otherwise the signature of `call` takes no
+ * arguments).
+ *
+ * Typing is inferred from `callback`s param, and from the rest params that
+ * the optional `reducer` function accepts. Use **explicit** types for these
+ * to ensure that everything _else_ is well-typed.
+ *
+ * Notice that this function constructs a funnel **object**, and does **not**
+ * execute anything when called. The returned object should be used to execute
+ * the funnel via the its `call` method.
+ *
+ * Debouncing: use `burstCoolDownMs` and any `invokedAt`.
+ * Throttling: use `delayMs` and `invokedAt: "start"` or `"both"`.
+ * Batching: See the reference implementation in [`funnel.reference-batch.test.ts`](https://github.com/remeda/remeda/blob/main/src/funnel.reference-batch.test.ts).
+ *
+ * @param callback - The main function that would be invoked occasionally based
+ * on `options`. The function would take the latest result of the `reducer`; if
+ * no calls where made since the last time it was invoked it will not be
+ * invoked. If a return value is needed, it should be passed via a reference or
+ * via closure to the outer scope of the funnel.
+ * @param options - An object that defines when `execute` should be
+ * invoked, relative to the calls of `call`. A timer that isn't defined will
+ * **not** be enabled, 0 is not used a default fallback.
+ * @param options.reducer - Combines the arguments passed to `call` with the
+ * value computed on the previous call (or `undefined` on the first time). The
+ * goal of the function is to extract and summarize the data needed for
+ * `callback`. It should be fast and simple as it is called often and should
+ * defer heavy operations to the `execute` function. If the final value
+ * is `undefined`, `callback` will not be called.
```

Comment:
**thought:** i can think of four common reducer cases:

1. no reducer — when the callback takes no params
2. use the first value — this should be, in theory, `identity()`, but that's not true because of the `undefined` the first time
3. use the last value — this is `(_, ...args) => args`, that's fine
4. array of all values — what the test calls `ARGS_COLLECTOR`.

we have a good solution for 1 now, maybe not the rest

Reply:
i guess the fact that we have an example for 4 is fine

Reply:
maybe we should have a wiki page for `funnel`

Reply:
Following a similar discussion we had around groupBy/countBy (e.g. "reduceBy") we can consider exporting a folder of "reduce utils" that would provide functions like below, 

```ts
const grouper = <T>(accum: readonly T[] | undefined, item: T) => [...(accum ?? []), item];
// e.g. groupBy = reduceBy(grouper);

const first = <T>(accum: T | undefined, item: T) => accum ?? item;

const last = <T>(accum: T | undefined, item: T) => item;

const counter = (accum: number | undefined, item: unknown) => (accum ?? 0) + 1
// e.g. countBy = reduceBy(counter);

const max = (accum: number | undefined, item: number) => Math.max(accum ?? -Infinity, item);

const min = (accum: number | undefined, item: number) => Math.min(accum ?? Infinity, item); 
```

Speaking more broadly, though, a reducer provides users the flexibility to use our functions as a well-tested base, which on top, they can build very custom logic, imagine for example, that you are debouncing mouse-movement events (which fire a lot) to log how far the user moved their mouse after clicking a button. Instead of adding all these event objects into an array so that you can process them when the function is invoked (to extract the max distance AND send the result to the server), you can build a custom reducer that, on every call to the funnel, extracts the distance from a single event, and check if that distance is the max. This means that instead of wasting all that time and memory building a huge array of events, you only maintain the information you need for the actual invocation.

I think that designing the API like this allows us to make both scenarios (the common "I just need the first/last item" and the rare "I need to keep this specific datum") easy to write and read.